crates/shim/src/io.rs

```diff
@@ -0,0 +1,289 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+extern crate libc;
+
+use std::fmt::Debug;
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Write};
+use std::os::unix::fs::OpenOptionsExt;
+use std::process::Command;
+use std::sync::Arc;
+use std::thread::JoinHandle;
+
+use crossbeam::sync::WaitGroup;
+use nix::sys::termios::Termios;
+
+use crate::error::{Error, Result};
+use crate::util::IntoOption;
+
+pub trait WriteCloser: Write {
+    fn close(&self);
```

Comment:
We generally should prefer `Drop` over `close`.

Reply:
Done